sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala

```diff
@@ -558,20 +560,27 @@ object ViewHelper {
       catalog: SessionCatalog, child: LogicalPlan): (Seq[Seq[String]], Seq[String]) = {
     def collectTempViews(child: LogicalPlan): Seq[Seq[String]] = {
       child.collect {
-        case UnresolvedRelation(nameParts, _, _) if catalog.isTempView(nameParts) =>
-          Seq(nameParts)
-        case plan if !plan.resolved => plan.expressions.flatMap(_.collect {
+        case s @ SubqueryAlias(_, view: View) if view.isTempView =>
+          Seq(s.identifier.qualifier :+ s.identifier.name)
+        case s: SubqueryAlias if s.getTagValue(SUBQUERY_TYPE_TAG).exists(_ == "tempView") =>
```

Comment:
This seems a bit hacky, but `SubqueryAlias(_, view: View)` couldn't handle all the cases. For example,
```scala
spark.range(10).createTempView("t")
sql("CREATE VIEW v AS SELECT * FROM t")
```
The `child` is:
```
Project [id#16L]
+- SubqueryAlias t
   +- Range (0, 10, step=1, splits=Some(2))
```
@cloud-fan, @viirya Any suggestion to capture the view properly? Does it make sense to add an `isView` field to `SubqueryAlias`?

Reply:
Or is it safe to do: `case s: SubqueryAlias if catalog.isTempView(s.identifier.name)`?

Reply:
This is a good point. We probably should use `View` to wrap temp view as well, to be truly consistent with permanent views. cc @linhongliu-db

Reply:
Thanks, will try that.

Reply:
+1 if it works, adding tag seems hacky.

Reply:
I tried to wrap it with `View`, but encountered an issue where duplicate columns are not allowed for `CatalogTable.properties`: https://github.com/apache/spark/blob/0592503669c27bf3140160f665d71ae388d342c3/sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala#L461-L464

Further looking, #30567 may have introduced a breaking change where the following works in Spark 3.0:
```
Seq(1).map(i => (i, i)).toDF("i", "i").createTempView("tv1")
sql("create temporary view tv2 as select 1 as i, 2 as i")
```
, but in Spark 3.1
```
Seq(1).map(i => (i, i)).toDF("i", "i").createTempView("tv1") // work
sql("create temporary view tv2 as select 1 as i, 2 as i") // doesn't work!
org.apache.spark.sql.AnalysisException: Found duplicate column(s) in the view definition: `i`
```

Should I fix this first such that the behavior is consistent with DataFrame API / DDL (basically we can relax the check only for temporary views)? WDYT @cloud-fan / @viirya ?

Reply:
I think it's expected. The goal is to make SQL temp view and permanent view consistent, and forbidding duplicated column names is consistent.

Reply:
`Seq(1).map(i => (i, i)).toDF("i", "i").createTempView("tv1")` is not SQL temp view (but dataframe temp view), so it's still allowed to have duplicated column names.

Reply:
Thanks @cloud-fan for the explanation.

I think we have few ways to detect dataframe temp views:
1. Add `isView: Boolean` field to the `SubqueryAlias`.
2. If we want to wrap dataframe temp views with `View`, we need to create `CatalogTable` - which is `View.desc` - with relaxed requirements (e.g., allowing duplicate columns).
3. Make `View.desc` optional and allow it to be `None` only for temporary views.

I think 3) makes the most sense without changing semantics.

Reply:
Shall we not wrap dataframe temp view with `View`? The original goal is to make SQL temp view and permanent view consistent.

Reply:
OK. In that case, we are left with two options:
1) Add `isView: Boolean` field to the `SubqueryAlias`, or
2) Introduce a new logical plan for handling dataframe temp view.

Reply:
I will try 2) if there is no opposition to the approach.

Reply:
OK I get your point now. If we need a new logical plan, then it's better to reuse `View` with some modification. +1 to make `View.desc` optional and set it to None for dataframe temp view. Or we can generate column names for dataframe temp view in `View.desc`, like `col1, col2, ...`, to bypass the check.

Reply:
@cloud-fan / @viirya I updated to make `View.desc` optional to support dataframe temp views. Could you please check again? Thanks!